google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessage.java

```diff
@@ -0,0 +1,328 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigquery.storage.v1alpha2;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.Message;
+import com.google.protobuf.UninitializedMessageException;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Converts Json data to protocol buffer messages given the protocol buffer descriptor. The protobuf
+ * descriptor must have all fields lowercased.
+ */
+public class JsonToProtoMessage {
+  private static ImmutableMap<FieldDescriptor.Type, String> FieldTypeToDebugMessage =
+      new ImmutableMap.Builder<FieldDescriptor.Type, String>()
+          .put(FieldDescriptor.Type.BOOL, "boolean")
+          .put(FieldDescriptor.Type.BYTES, "string")
+          .put(FieldDescriptor.Type.INT32, "int32")
+          .put(FieldDescriptor.Type.DOUBLE, "double")
+          .put(FieldDescriptor.Type.INT64, "int64")
+          .put(FieldDescriptor.Type.STRING, "string")
+          .put(FieldDescriptor.Type.MESSAGE, "object")
+          .build();
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  public static DynamicMessage convertJsonToProtoMessage(
+      Descriptor protoSchema, JSONObject json, boolean allowUnknownFields)
+      throws IllegalArgumentException {
+    if (json.length() == 0) {
+      throw new IllegalArgumentException("JSONObject is empty.");
+    }
+    return convertJsonToProtoMessageImpl(
+        protoSchema, json, "root", /*topLevel=*/ true, allowUnknownFields);
+  }
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @param jsonScope Debugging purposes
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @param topLevel checks if root level has any matching fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  private static DynamicMessage convertJsonToProtoMessageImpl(
+      Descriptor protoSchema,
+      JSONObject json,
+      String jsonScope,
+      boolean topLevel,
+      boolean allowUnknownFields)
+      throws IllegalArgumentException {
+
+    DynamicMessage.Builder protoMsg = DynamicMessage.newBuilder(protoSchema);
+    String[] jsonNames = JSONObject.getNames(json);
+    int matchedFields = 0;
+    // JsonNames will be null if the JSONObject is empty.
+    if (jsonNames != null) {
```

Comment:
Return on null earlier so the flow is easier to understand.
if (jsonNames == null) {
   return protoMsg.build();
}